### DISCUSSION TITLE
Having different edit settings when creating in BagPart

### DISCUSSION THREAD

**Discussion by Psichorex** [OrchardCMS/OrchardCore#11044] [Q&A] [answered]
I am wondering that is it possible to have differend Edit.cshtml to run when creating a ContentItem from the New menu vs creating an item in a BagPart.
My goal is to have a detailed creating page when you create the item from admin menu.
But when I create the same type of item in a BagPart I want to have only the Title/Name to be available.
I'm new to Orchard and I can't really grap the point where the application directs you to the edit.cshtml page and it is frustrating as it seems to me that no matter from where you create an item the same .edit.cshtml will run

**Comment by Skrypt** [comment]
The Edit.cshtml view will render a BuildEditorAsync() method which will create the form required for that ContentItem. Then, the controller associated with this view expects to receive a Shape corresponding to that. So, the best way to hide fields is to use Javascript with specific content types. Not sure 100% how this would work with a BagPart. Maybe explain more in detail what you are trying to achieve and we can maybe find alternative solutions.

**Reply by Psichorex** [reply]
Let's Say there is a Content Type called Master
Master has
Name
Age
And a content picker Field to choose his trainee
I want that if I create a Master from admin UI I could set all 3 properties.
But when I have a Type for example MastersFacility with a Bagpart that contains masters if I add a new master to the MastersFacility I dont want to have the content picker Field to show up only the Name and Age.
My first approach was to somehow get the information where the creation is coming from so that with an If statement in the cshtml I could separate default creation from inside Bagpart creation

**Comment by Skrypt** [comment]
So here to resume we have : 

Master: Content Type with Name (Text Field), Age (Numeric Field), Trainee (Content Picker Field).
Trainee : Content Type with at least a TitlePart.

MastersFacility: Content Type that is containing a TitlePart and BagPart that allows containing Masters. This MastersFacility is creatable and probably has also an AutoroutePart so that it acts as a new page.

When you create a Master you want to be able to set Name, Age and Trainee from the Content List admin UI.
But when you create a MastersFacility "page" you want that the BagPart editor prevents you to assign a Trainee (Content Picker) to the Master you are adding.

Let me try this and I will report what I've found later on.

**Reply by Skrypt** [reply]
So basically we have this : 

![image](https://user-images.githubusercontent.com/3228637/150448452-0fba6c9b-2abf-4053-8ef5-3aae804f3a29.png)

But I'm wondering why you wouldn't want to let people assign a Trainee to the Master in that case.
Is it because of some role/permission condition?

**Reply by Skrypt** [reply]
So from my point of view here we still are using a BuildEditorAsync() method to create these form elements to be able to add a Master in the BagPart. So, the only way is to hide this ContentPicker using javascript. So as you can see here in that screenshot : 

![image](https://user-images.githubusercontent.com/3228637/150448994-7ff9cee3-9fc0-4174-9c0d-b5d11a8ed2ee.png)

JQuery selector would be something like : 

`$('#BagPart_widgetTemplatePlaceholder #ContentPicker_Master_Trainee_').css('display', 'none');`

You could add this in a "Master" Content Type editor shape template.

Here, I'm saying that you should use javascript to only hide the form elements because else the Driver expects receiving all these form elements to create a proper master still. So, you can just remove the form elements from DOM from javascript either.

**Reply by Skrypt** [reply]
I think we are missing that shape template extensibility but I need to double-check. Alternative would be to try adding this directly in an override of the Edit.cshtml file you are referring to.

**Reply by Psichorex** [reply]
Yes you got my problem absolutely right. I want this feature because I want the BagPart creation to be a less detailed version.
I could have said any property instead of the content picker field to hide but that was an example. The point is that when in BagPart the creation properties should be limited compared to the normal creation.
I will check the javascript version. 



> I think we are missing that shape template extensibility but I need to double-check. Alternative would be to try adding this directly in an override of the Edit.cshtml file you are referring to.

Yes that was my problem. So whenever you click on new Master either form admin UI or from BagPart it will call the Edit method of the DisplayDriver which will first populate the Master entity and then it will direct you to the corressponding MasterPart.Edit.cshtml
In this case you have no mapping from where have you reached this MasterPart.Edit.cshtml so you can't differentiate between admin UI create and BagPart create. So probably your JavaScript solution will be the one

**Reply by deanmarcussen** [reply]
I think this is probably a duplicate of this issue https://github.com/OrchardCMS/OrchardCore/issues/6784

Happy too see a pr with alternates.

**Reply by Psichorex** [reply]
> I think this is probably a duplicate of this issue #6784
> 
> Happy too see a pr with alternates.

I have read the issue you have mentioned but I don't quite get whether you reached a solution there or not.

**Reply by Psichorex** [reply]
@Skrypt To ensure that you got the problem right here is my approach
So here is master creation from admin UI new

![masterfromadmin](https://user-images.githubusercontent.com/92299130/150490666-6e3a63a4-632c-4a10-957d-37e14b4ca7f2.PNG)

Here is Master creation from Bag as I would want it 
![masterfrombag](https://user-images.githubusercontent.com/92299130/150490793-b5b2d70b-4f90-4ae8-a147-d26e3b7378f1.PNG)

**Comment by ns8482e** [comment]
@Psichorex same answer as in https://github.com/OrchardCMS/OrchardCore/issues/11046#issuecomment-1018837414

Create template `ContentCard_Frame__MasterFacility__BagPart__Master` (`ContentCard-MasterFacility-BagPart-Master.Frame.cshtml`)  and remove the shapes from editor
```cshtml
@{
// Remove Field shape in case of BagPart
Model.ChildContent.ContentEditor.Parts.MasterPart.Remove("MasterPart-Trainee")
}
<div class="widget-template col-md-12 px-1">
    @await DisplayAsync(Model.ChildContent)
</div>
```

**Reply by Psichorex** [reply]
> @Psichorex same answer as in https://github.com/OrchardCMS/OrchardCore/issues/11046#issuecomment-1018837414
> 
> Create template `ContentCard_Frame__MasterFacility__BagPart__Master` (`ContentCard-MasterFacility-BagPart-Master.Frame.cshtml`)  and remove the shapes from editor
> ```cshtml
> @{
> // Remove Field shape in case of BagPart
> Model.ChildContent.ContentEditor.Parts.MasterPart.Remove("MasterPart-Trainee")
> }
> <div class="widget-template col-md-12 px-1">
>     @await DisplayAsync(Model.ChildContent)
> </div>
> ```

So I create a `ContentCard-MasterFacility-BagPart-Master.Frame.cshtml` in which I implement that code part you posted and that should disallow the rendering of the content Field?

**Reply by ns8482e** [reply]
yes that's the alternate (`ContentCard_Frame__[ParentContentType]__[CollectionType]__[ContentType]`) for `ContentCard_Frame` that represents outer frame of each element in BagPart 

Within that template `Model.ChildContent` is `ContentCard` shape and `Model.ChildContent.ContentEditor` is the content item editor shape which is same shape as in via Edit.cshtml 

In content editor shape,  all parts are placed in `Parts` zone 
Fields are placed inside each part.

so by following line
`Model.ChildContent.ContentEditor.Parts.MasterPart.Remove("MasterPart-Trainee")`

you are removing Trainee field shape from MasterPart shape of the editor  shape.

you can use console_log  on Model  to check how shapes are outlined.

**Comment by Psichorex** [comment, marked as answer]
My probelm was solved using IShapeTableProvider
I searched for a difference between two of the same content item if one is inside a BagPart
Turned out the their Prefix is different.
Using this difference I was able to make the ContentPickerField of the BagPart contained item disappear

```
public void Discover(ShapeTableBuilder builder)
        {
            builder.Describe("ContentPickerField_Edit")
                .OnDisplaying(displaying =>
                {
                    var editor = displaying.Shape;                                                         
                    if (differentiator logic e.g. prefix difference)
                    {
                        editor.Metadata.ChildContent = HtmlString.Empty;
                    }
                });
        }
```
